### PR TITLE
Added HAL modulus component.

### PR DESCRIPTION
**Pull request by propcoder**

### CODE REVIEW

**Comment by mhaberler** [comment]
congratulations! you took the dive - super

the buildbot is unhappy but that does not mean it is this comp - it looks a bit like general indigestion

**Comment by zultron** [comment]
```
make[2]: *** No rule to make target 'objects/posix/halcomp-srcs/hal/i_components/modulus.o', needed by '../rtlib/posix/modulus.so'.  Stop.
make[2]: *** Waiting for unfinished jobs....
```
(repeated for other userland flavors)

Also a compile warning in kthreads:
```
hal/components/modulus.comp: In function '_':
hal/components/modulus.comp:13:24: warning: 'return' with a value, in function returning void [enabled by default]
hal/components/modulus.comp:19:5: warning: 'return' with a value, in function returning void [enabled by default]
```

**Comment by strahlex** [comment]
You do not need to create a non-instantiable .comp. The icomp is enough.

File: src/hal/components/modulus.comp [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+pin in float divisor "Divisor to be divided by dividend";
+pin in float dividend = 1.0 "Dividend";
+pin out float out "Modulus output";
+
+function _ fp;
+license "GPLv2 or later";
+;;
+#include <rtapi_math.h>
+
+#define _INFINITY (1.0 / 0.0)
```
Comment (strahlex) [inline src/hal/components/modulus.comp]: There should be an INFINITY macro. Can you please check.

File: src/hal/i_components/modulus.icomp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+pin in float divisor "Divisor to be divided by dividend";
+pin in float dividend = 1.0 "Dividend";
+pin out float out "Modulus output";
+
+function _;
+license "GPL";
+;;
+#include <rtapi_math.h>
+
+#define _INFINITY (1.0 / 0.0)
```
Comment (strahlex) [inline src/hal/i_components/modulus.icomp]: Same here.

File: src/hal/i_components/modulus.icomp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+function _;
+license "GPL";
+;;
+#include <rtapi_math.h>
+
+#define _INFINITY (1.0 / 0.0) 
+
+FUNCTION(_) {
+    if (dividend == 0) {
```
Comment (strahlex) [inline src/hal/i_components/modulus.icomp]: Use 0.0 to indicate a float comparison.

File: src/hal/i_components/modulus.icomp [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        out = _INFINITY;
+        return 0;
+    }
+    hal_float_t tmp;
+    tmp = divisor / dividend;
+    if (tmp < 0) out = divisor + rtapi_floor(1 - tmp) * dividend;
```
Comment (strahlex) [inline src/hal/i_components/modulus.icomp]: Please use braces. Also use 1.0 to indicate that the number is a float value.

File: src/hal/i_components/modulus.icomp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    else out = divisor - rtapi_floor(tmp) * dividend;
```
Comment (strahlex) [inline src/hal/i_components/modulus.icomp]: Please use braces.

**Comment by strahlex** [comment]
@propcoder Overal good work. Please look at my comments regarding the code style.

**Comment by mhaberler** [comment]
While perusing hal/components I found this component which has no man page but obviously targets a similar problem: 

https://github.com/machinekit/machinekit/blob/master/src%2Fhal%2Fcomponents%2Fmodmath.c

this might be worth studying and documenting - could well be the solution is already in place

**Comment by mhaberler** [comment]
anything forthcoming here?

**Comment by propcoder** [comment]
No, nothing.

And git.. it is still a very rarely needed (if at all) tool for me. I don't
let myself to struggle learning git while I have 10 more subjects to do.
This is now.

2015-10-08 21:03 GMT+03:00 Michael Haberler <notifications@github.com>:

> anything forthcoming here?
>

**Comment by mhaberler** [comment]
can be reopened in case determination reappears.

**Pull request closed without merging**